frontend/model/contracts/chatroom.js

```diff
@@ -0,0 +1,370 @@
+  type: string, contractID: string, hash: string, state: Object
+}): void {
+  for (let i = state.messages.length - 1; i >= 0; i--) {
+    if (state.messages[i].id === hash) {
+      sbp('okTurtles.events/emit', `${CHATROOM_MESSAGE_ACTION}-${contractID}`, {
+        type,
+        data: { message: state.messages[i] }
+      })
+      break
+    }
+  }
+}
+
+sbp('chelonia/defineContract', {
+  name: 'gi.contracts/chatroom',
+  metadata: {
+    validate: objectOf({
+      createdDate: string, // action created date
+      username: string, // action creator
+      identityContractID: string // action creator identityContractID
+    }),
+    create () {
+      const { username, identityContractID } = sbp('state/vuex/state').loggedIn
+      return {
+        createdDate: new Date().toISOString(),
+        username,
+        identityContractID
+      }
+    }
+  },
+  state (contractID) {
+    return sbp('state/vuex/state')[contractID]
+  },
+  getters: {
+    currentChatRoomState (state) {
+      return state
+    },
+    chatRoomSettings (state, getters) {
+      return getters.currentChatRoomState.settings || {}
+    },
+    chatRoomAttributes (state, getters) {
+      return getters.currentChatRoomState.attributes || {}
+    },
+    chatRoomUsers (state, getters) {
+      return getters.currentChatRoomState.users || {}
+    },
+    chatRoomLatestMessages (state, getters) {
+      const messages = getters.currentChatRoomState.messages || []
+      return getLatestMessages({
+        count: getters.chatRoomSettings.messagesPerPage,
+        messages
+      })
+    }
+  },
+  actions: {
+    // This is the constructor of Chat contract
+    'gi.contracts/chatroom': {
+      validate: objectOf({
+        attributes: chatRoomAttributesType
+      }),
+      process ({ meta, data }, { state }) {
+        const initialState = merge({
+          settings: {
+            messagesPerPage: CHATROOM_MESSAGES_PER_PAGE,
+            maxNameLength: CHATROOM_NAME_LIMITS_IN_CHARS,
+            maxDescriptionLength: CHATROOM_DESCRIPTION_LIMITS_IN_CHARS
+          },
+          attributes: {
+            creator: meta.username,
+            deletedDate: null,
+            archievedDate: null
+          },
+          users: {},
+          messages: []
+        }, data)
+        for (const key in initialState) {
+          Vue.set(state, key, initialState[key])
+        }
+      },
+      sideEffect ({ contractID, meta }) {
+        const rootState = sbp('state/vuex/state')
+        // this condition makes users redirect to the newly-created channel only after they created a new channel
+        if (meta.username === rootState.loggedIn.username) {
+          sbp('okTurtles.events/once', CONTRACT_IS_SYNCING, (cID, isSyncing) => {
+            if (cID === contractID && !isSyncing) {
+              sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
+            }
+          })
+        }
+      }
```

Comment:
I don't think this is the right place to set `setCurrentChatRoomId`.

It makes an assumption that we should set this to be the current chatroom based on whether or not `meta.username === rootState.loggedIn.username`, and then it creates an event listener... This really doesn't look correct to me. We're just processing the beginning of a chatroom contract, and it doesn't seem right to me that we should have any sideEffects here just because we've subscribed to a chatroom.

I think the more correct place would be to `setCurrentChatRoomId` would be in one of the `gi.actions`, either in `controller/actions/group.js` or in `controller/actions/chatroom.js`, after we've done `syncAndWait`.

Reply:
> It makes an assumption that we should set this to be the current chatroom based on whether or not meta.username === rootState.loggedIn.username, and then it creates an event listener... This really doesn't look correct to me. We're just processing the beginning of a chatroom contract, and it doesn't seem right to me that we should have any sideEffects here just because we've subscribed to a chatroom.

Since `meta.username` is the chatroom creator, only creator will listen the event `CONTRACT_IS_SYNCING` once.
The time of setting this event is when he started to sync but not finished syncing.
I don't think there will be any problems... 🙂

> I think the more correct place would be to setCurrentChatRoomId would be in one of the gi.actions, either in controller/actions/group.js or in controller/actions/chatroom.js, after we've done syncAndWait.

I don't think we have any other place to call `setCurentChatRoomId`. That's because We call `syncAndWait` function only in `sideEffect` of group action `joinChatRoom` and we don't use this `joinChatRoom` only after creating a new chatroom. We use this `joinChatRoom` also for just joining chatroom too.

Reply:
This sideEffect really doesn't make sense being here.

This is the sideEffect for `'gi.contracts/chatroom'`. What does `'gi.contracts/chatroom'` do?

It creates / registers a chatroom. It doesn't have anything to do with setting the current chatroom, so it doesn't belong here.

Setting the current chatroom should be paired with a user action.

> I don't think we have any other place to call `setCurentChatRoomId`

You can do it inside of `'gi.actions/group/joinChatRoom'` - so that when the user joins a chatroom, they also set the current chatroom id.

Reply:
Do you think it should be `gi.actions/group/addAndJoinChatRoom` or `gi.actions/group/addChatRoom`?

```
sbp('okTurtles.events/once', CONTRACT_IS_SYNCING, (cID, isSyncing) => {
  if (cID === contractID && !isSyncing) {
    sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: contractID })
  }
})
```

Reply:
I think it should be htere inside this `gi.actions/group/addAndJoinChatRoom`.  :grin:

Reply:
@Silver-IT why in `gi.actions/group/addAndJoinChatRoom`? What if they're not adding the channel, but are another user joining and switching to that channel?

Then I don't think it will work, because the channel already exists in the group. So I think it should go into `'gi.actions/group/joinChatRoom'`

Reply:
The following is the updated one. What do you think?
```
'gi.actions/group/addAndJoinChatRoom': async function (params: GIActionParams) {
    const message = await sbp('gi.actions/group/addChatRoom', {
      ...omit(params, ['options']),
      hooks: {
        prepublish: params.hooks?.prepublish,
        postpublish: null
      }
    })

    const joiningMessage = await sbp('gi.actions/group/joinChatRoom', {
      ...omit(params, ['options']),
      data: {
        chatRoomID: message.contractID()
      },
      hooks: {
        prepublish: null,
        postpublish: params.hooks?.postpublish
      }
    })

    sbp('okTurtles.events/once', joiningMessage.hash(), (cId, msg) => {
      sbp('state/vuex/commit', 'setCurrentChatRoomId', { chatRoomId: cId })
    })

    return message
  },
```

Reply:
The current workflow is like that `If user creates a new chatroom, he will be automatically redirected to the chatroom`.
So I think we should set `currentChatRoomId` in `addChatRoom` or `addAndJoinChatRoom`.

Reply:
@Silver-IT They should set the `setCurrentChatRoomId` when they join a chatroom even if they didn't create it right?

If so, then it should be added to `'gi.actions/group/joinChatRoom'`, and not to `addChatRoom`.

Reply:
> @Silver-IT They should set the setCurrentChatRoomId when they join a chatroom even if they didn't create it right?

No. That's because UserA can add UserB to the channel. In this case, UserB shouldn't be redirected to the new channel unexpectedly.

Reply:
@Silver-IT but what about when UserA adds themselves to a channel that they didn't create?

Reply:
I don't understand your question. If user creates a new channel, he will be redirected to the channel automatically.
That's what we are going to do.

If UserA adds UserB to the channel (which UserA has created), then in UserB, nothing happens except joining in the background. No switching channel here.

Reply:
I'll reply on Slack